plutus-benchmark/bls12-381-costs/src/PlutusBenchmark/BLS12_381/Scripts.hs

```diff
@@ -289,20 +286,608 @@ groth16Verify (Tx.bls12_381_G1_uncompress -> alpha')
+
+    -- define the third and last element of a proof of correct VRF
+    s = (k - (os2ip c) * privKey) `modulo` 52435875175126190479447740508185965837690552500527637822603658699938581184513
+
+    -- Define the cofactor of G2
+    f = 305502333931268344200999753193121504214466019254188142667664032982267604182971884026507427359259977847832272839041692990889188039904403802465579155252111 :: Integer
+
+    -- create our VRF hash output
+    beta = sha2_256 . Tx.bls12_381_G2_compress $ Tx.bls12_381_G2_scalarMul f gamma
+
+    -- The proof of that the VRF hash of input alpha under our priv key is beta
+    pi = (gamma, c, s)
+
+    -- To verify a VRF hash given an
+    --        input alpha
+    --        output beta
+    --        proof pi (gamma, c, s)
+    --        pubkey pub
+    -- do the following calculation
+    u  = Tx.bls12_381_G2_add (Tx.bls12_381_G2_scalarMul (os2ip c) pub) (Tx.bls12_381_G2_scalarMul s g2Gen)
+    h' = Tx.bls12_381_G2_hashToGroup message emptyByteString
+    v  = Tx.bls12_381_G2_add (Tx.bls12_381_G2_scalarMul (os2ip c) gamma) (Tx.bls12_381_G2_scalarMul s h')
+
+  -- and check
+  c == (sha2_256 . mconcat $ Tx.bls12_381_G2_compress <$> [g2Gen,h',pub,gamma,u,v])
+
+  where
+    os2ip :: BuiltinByteString -> Integer
+    os2ip bs
+      | bs == emptyByteString = error ()
+      | otherwise             = go bs
+      where
+        len xs = lengthOfByteString xs - 1
+        intAtLastByte xs = indexByteString xs $ len xs
+        stripLastByte xs = takeByteString (len xs) xs
+        go xs
+          | xs == emptyByteString = 0
+          | otherwise             = intAtLastByte xs + 256 * go (stripLastByte xs)
+
+checkVrfBlsScript :: Bool
+checkVrfBlsScript = vrfBlsScript vrfPrivKey vrfMessage g2Generator
+
+mkVrfBlsPolicy :: UPLC.Program UPLC.NamedDeBruijn DefaultUni DefaultFun ()
+mkVrfBlsPolicy =
+    Tx.getPlcNoAnn $ $$(Tx.compile [|| vrfBlsScript ||])
+      `Tx.unsafeApplyCode` Tx.liftCodeDef vrfPrivKey
+      `Tx.unsafeApplyCode` Tx.liftCodeDef vrfMessage
+      `Tx.unsafeApplyCode` Tx.liftCodeDef g2Generator
+
+---------------- Verify over G1 ----------------
+
+{- BLS signature with the public key over G1. This function returns a message `msg`, a public
+   key `pk`, and a signature `sig`. Verification of these test vectors should proceed as follows:
+    * pk_deser = G1Decompress(pk)
+    * sig_deser = G2Decompress(sig)
+    * hashed_msg = G2HashToCurve(msg, "BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_")
+    * Check that pairing(pk_deser, hashed_msg) = pairing(G1Generator, sig_deser)
+-}
+
+{-# INLINABLE g1VerifyMessage #-}
+g1VerifyMessage :: BuiltinByteString
+g1VerifyMessage  = toBuiltin $ bytesFromHex "3e00ef2f895f40d67f5bb8e81f09a5a12c840ec3ce9a7f3b181be188ef711a1e"
+
+{-# INLINABLE g1VerifyPubKey #-}
+g1VerifyPubKey :: BuiltinByteString
+g1VerifyPubKey = toBuiltin $ bytesFromHex ("aa04a34d4db073e41505ebb84eee16c0094fde9fa22ec974" <>
```

Comment:
Where did the test vectors (and indeed the examples themselves) come from?  If they're from a standard or something it might be worth saying that.